src/main/java/org/opensearch/timeseries/rest/handler/IntervalCalculation.java

```diff
@@ -216,109 +228,535 @@ public void onFailure(Exception e) {
+    public void findMedianIntervalAdaptive(ActionListener<IntervalTimeConfiguration> listener) {
+
+        final long MIN_BUCKET_WIDTH_MINS = 1;
+        final String TS_FIELD = config.getTimeField();
+
+        /*
+         * --------------------------------------------------------------
+         * the common filter (time-range + caller predicates) *
+         * --------------------------------------------------------------
+         */
+        BoolQueryBuilder baseFilter = QueryBuilders.boolQuery().filter(config.getFilterQuery());
+
+        if (topEntity != null && !topEntity.isEmpty()) {
+            Entity e = Entity.createEntityByReordering(topEntity);
+            for (TermQueryBuilder tq : e.getTermQueryForCustomerIndex()) {
+                baseFilter.filter(tq);
+            }
+        }
+
+        /*
+         * --------------------------------------------------------------
+         * request: get min / max timestamp + total hits *
+         *
+         * "bounds" query built by boundsSrc
+         *
+         * (Everything under `"filter": […]` comes from     baseFilter.
+         *  It already includes customer predicates and,            if
+         *  present, the topEntity terms. No date-range yet — we are
+         *  discovering min/max here.)
+         *
+         *  --------------------------------------------------------------------
+         {
+            "size": 0,
+            "query": {
+                "bool": {
+                    "filter": [
+                        {
+                            "match_all": {
+                                "boost": 1
+                            }
+                        }
+                    ],
+                    "adjust_pure_negative": true,
+                    "boost": 1
+                }
+            },
+            "aggregations": {
+                "min_ts": {
+                    "min": {
+                        "field": "timestamp"
+                    }
+                },
+                "max_ts": {
+                    "max": {
+                        "field": "timestamp"
+                    }
+                }
+            }
+        }
+         * --------------------------------------------------------------
+         */
+        SearchSourceBuilder boundsSrc = new SearchSourceBuilder()
+            .size(0)
+            .trackTotalHits(true)
+            .query(baseFilter)
+            .aggregation(AggregationBuilders.min("min_ts").field(TS_FIELD))
+            .aggregation(AggregationBuilders.max("max_ts").field(TS_FIELD));
+
+        SearchRequest boundsReq = new SearchRequest(config.getIndices().toArray(new String[0])).source(boundsSrc);
+
+        client.search(boundsReq, ActionListener.wrap(r -> {
+
+            Min minAgg = r.getAggregations().get("min_ts");
+            Max maxAgg = r.getAggregations().get("max_ts");
+
+            if (minAgg == null
+                || maxAgg == null
+                || Double.isInfinite(minAgg.getValue())
```

Comment:
should we add NaN check here also, not sure if aggregation can ever return NaN here or only if we are dividing by zero which doesn't happen here.

Reply:
added